app/lib/features/attachments/widgets/reference_attachment_item.dart

```diff
@@ -25,74 +35,178 @@ class ReferenceAttachmentItem extends ConsumerWidget {
-          ],
-        ),
+        onTap: () => onTapRefAttachment(context, ref),
+        leading: refAttachmentIcons(ref),
+        title: refAttachmentTitle(ref),
+        subtitle: refAttachmentSubTitle(),
+        trailing: actionMenu(context),
       ),
     );
   }
 
-  Widget title(BuildContext context) {
-    final title = refDetails.title();
-    return Column(
-      crossAxisAlignment: CrossAxisAlignment.start,
+  Widget refAttachmentTitle(WidgetRef ref) {
+    final defaultTitle = Text(refDetails.title().toString());
+    final refObjectType = refDetails.typeStr();
+    final refObjectId = refDetails.targetIdStr();
+
+    if (refObjectId == null) return defaultTitle;
+    switch (refObjectType) {
+      case 'pin':
+        final pin = ref.watch(pinProvider(refObjectId)).valueOrNull;
+        return pin == null ? defaultTitle : Text(pin.title().toString());
+      case 'calendar-event':
+        final event = ref.watch(calendarEventProvider(refObjectId)).valueOrNull;
+        return event == null ? defaultTitle : Text(event.title().toString());
+      case 'task-list':
+        final taskList = ref.watch(taskListProvider(refObjectId)).valueOrNull;
+        return taskList == null
+            ? defaultTitle
+            : Text(taskList.name().toString());
+      default:
+        return defaultTitle;
+    }
+  }
+
+  Widget refAttachmentSubTitle() {
+    return Text(refDetails.typeStr().toString());
+  }
+
+  Widget refAttachmentIcons(WidgetRef ref) {
+    final defaultIcon = PhosphorIcon(PhosphorIconsThin.tagChevron, size: 30);
+    final refObjectType = refDetails.typeStr();
+    final refObjectId = refDetails.targetIdStr();
+
+    if (refObjectId == null) return defaultIcon;
+    switch (refObjectType) {
+      case 'pin':
+        final pin = ref.watch(pinProvider(refObjectId)).valueOrNull;
+        if (pin == null) return defaultIcon;
+        return ActerIconWidget(
+          iconSize: 30,
+          color: convertColor(
+            pin.display()?.color(),
+            iconPickerColors[0],
+          ),
+          icon: ActerIcon.iconForPin(
+            pin.display()?.iconStr(),
+          ),
+        );
+      case 'calendar-event':
+        return PhosphorIcon(PhosphorIconsThin.calendar, size: 30);
+      case 'task-list':
+        final taskList = ref.watch(taskListProvider(refObjectId)).valueOrNull;
+        if (taskList == null) return defaultIcon;
+        return ActerIconWidget(
+          iconSize: 30,
+          color: convertColor(
+            taskList.display()?.color(),
+            iconPickerColors[0],
+          ),
+          icon: ActerIcon.iconForPin(
+            taskList.display()?.iconStr(),
+          ),
+        );
+      default:
+        return defaultIcon;
+    }
+  }
+
+  void onTapRefAttachment(
+    BuildContext context,
+    WidgetRef ref,
+  ) {
+    final lang = L10n.of(context);
+    final refObjectType = refDetails.typeStr();
+    final refObjectId = refDetails.targetIdStr();
```

Comment:
having several functions all call the exact same details feels wrong and in particular that they then have matches on them...

I'd rather say, let's make separate Widgets for each type (`pin`, `calendar-event`, etc) and have one global match that selects the right widget and passes the ref-details over..

makes it also a lot neater and easier to extend than extending several functions with their internal match ...

Reply:
Yes agree with you and also felt the same during this code development. Tried couple of way but then reverted it to minimised this PR Complexity. 

Surely will do this in next follow-up PR.